### ISSUE TITLE
map_entry, suggestion produces invalid borrow inside closure

### ISSUE THREAD

**Issue by Lunderberg** [rust-lang/rust-clippy#11976]
### Summary

The `map_entry` lint is suppressed if the hash map is accessed between the call to `HashMap::contains_key` and the later call to `HashMap::insert`.  However, it does not check for access of the hash map occurring within a closure.  As a result, the suggestion's use of `HashMap::entry` starts a mutable borrow of the entire hashmap, conflicting with call to `HashMap::get` that occurred between the `contains_key` and `insert`.

In the example below, if the line `let _ = || hashmap.get(&0);` is replaced with `hashmap.get(&0);`, the lint is no longer suggested.

### Lint Name

map_entry

### Reproducer

I tried this code:

```rust
fn main() {
    let mut hashmap = std::collections::HashMap::new();
    if !hashmap.contains_key(&0) {
        let _ = || hashmap.get(&0);
        hashmap.insert(0, 0);
    }
}
```

I saw this happen:

```
warning: usage of `contains_key` followed by `insert` on a `HashMap`
 --> src/main.rs:3:5
  |
3 | /     if !hashmap.contains_key(&0) {
4 | |         let _ = || hashmap.get(&0);
5 | |         hashmap.insert(0, 0);
6 | |     }
  | |_____^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_entry
  = note: `#[warn(clippy::map_entry)]` on by default
help: try
  |
3 ~     hashmap.entry(0).or_insert_with(|| {
4 +         let _ = || hashmap.get(&0);
5 +         0
6 +     });
  |

warning: `temp` (bin "temp") generated 1 warning (run `cargo clippy --fix --bin "temp"` to apply 1 suggestion)
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
```

I expected to see this happen: No clippy suggestions produced.

### Version

```text
rustc 1.76.0-nightly (06e02d5b2 2023-12-09)
binary: rustc
commit-hash: 06e02d5b259c1e88cbf0c74366d9e0a4c7cfd6d9
commit-date: 2023-12-09
host: x86_64-unknown-linux-gnu
release: 1.76.0-nightly
LLVM version: 17.0.5
```

**Comment by dhilipsiva** [comment]
I faced the same Issue. Here is the relevant PR: https://github.com/8xFF/atm0s-media-server/pull/349/files#diff-33ab08f3a3309009785786b7440abbb7e1f2a4ae26dc63c54f2d10acf16a2f56R68

**Comment by dhilipsiva** [comment]
More context from the above PR:

```
error[E0499]: cannot borrow `*self` as mutable more than once at a time
  --> packages/media_record/src/media.rs:69:34
   |
68 |                 if let std::collections::hash_map::Entry::Vacant(e) = self.tracks.entry(id) {
   |                                                                       ----------- first mutable borrow occurs here
69 |                     let writer = self.get_free_writer_for(event.ts, media.meta.is_audio());
   |                                  ^^^^ second mutable borrow occurs here
...
76 |                     e.insert(TrackWriter { writer });
   |                     - first borrow later used here

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0499`.
Original diagnostics will follow.

warning: usage of `contains_key` followed by `insert` on a `HashMap`
  --> packages/media_record/src/media.rs:68:17
   |
68 | /                 if !self.tracks.contains_key(&id) {
69 | |                     let writer = self.get_free_writer_for(event.ts, media.meta.is_audio());
70 | |                     if media.meta.is_audio() {
71 | |                         self.writers[writer].audio_inuse = true;
...  |
76 | |                     self.tracks.insert(id, TrackWriter { writer });
77 | |                 }
   | |_________________^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_entry
   = note: `#[warn(clippy::map_entry)]` on by default
help: try
   |
68 ~                 if let std::collections::hash_map::Entry::Vacant(e) = self.tracks.entry(id) {
69 +                     let writer = self.get_free_writer_for(event.ts, media.meta.is_audio());
70 +                     if media.meta.is_audio() {
71 +                         self.writers[writer].audio_inuse = true;
72 +                     } else {
73 +                         self.writers[writer].video_inuse = true;
74 +                     }
75 +                     log::info!("write track {:?} to writer {writer}", id);
76 +                     e.insert(TrackWriter { writer });
77 +                 }
   |

warning: `media-server-record` (lib) generated 1 warning (run `cargo clippy --fix --lib -p media-server-record` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 6.92s
```
